Enumerable methods ready for review

ahmetbozaci opened the pull request:
[Live Demo ](https://repl.it/@ahmt9417/EnumerableMethods-2#enumerables.rb)

Dear TSE,

As a project requirements I created **three branch.**
1. Master
2. Development (default branch)
3. Feature(working branch)

I worked the solo in this project, so I didn't change the branch when I worked on it. 

**I disabled some rubocop's properties.**
 
1. ModuleLength
2. CyclomaticComplexity,
3. PerceivedComplexity,
4. MethodLength

Because my module and my some methods (**my_inject**, **my_all**, **my_none**, **my_any**) has many lines. 
But I still refactored all methods and make them less line as I can do.

ahmetbozaci committed:
added rubocop

ahmetbozaci committed:
created my_each

ahmetbozaci committed:
created my_each_with_index

ahmetbozaci committed:
created my_map

ahmetbozaci committed:
added examples

ahmetbozaci committed:
added my_any

ahmetbozaci committed:
added my_select method

ahmetbozaci committed:
changed all each method with my_each in  my_any method

ahmetbozaci committed:
added my_all

ahmetbozaci committed:
added my_none

ahmetbozaci committed:
added examples1

ahmetbozaci committed:
added my_count method

ahmetbozaci committed:
modify my_map for proc

ahmetbozaci committed:
fixed my_each_With_index for range

ahmetbozaci committed:
fixed my_any for range

ahmetbozaci committed:
fixed my_all if no block given

ahmetbozaci committed:
fixed my_none for range

ahmetbozaci committed:
fixed my_count for range

ahmetbozaci committed:
added my_inject with examples

ahmetbozaci committed:
added multiply_els

ahmetbozaci committed:
fixed rubocop line errors

ahmetbozaci committed:
fixed rubocop self errors

ahmetbozaci committed:
fixed rubocop line errors

ahmetbozaci committed:
refactor my_map,my_select,my_each_with_index

ahmetbozaci committed:
refactor my_each method

ahmetbozaci committed:
refactor my_any,

ahmetbozaci committed:
refactor my_all

ahmetbozaci committed:
refactor my_none

ahmetbozaci committed:
refactor my_count

ahmetbozaci committed:
refactor my_inject

ahmetbozaci committed:
added readme

ahmetbozaci committed:
fixed repli link in readme file

ahmetbozaci committed:
Update README.md

ahmetbozaci committed:
Update README.md

andresporras3423 reviewed (changes requested):
### Status: Changes Required :recycle:
**Hello Ahmet Bozacı** 
Good job with your first submission  :muscle:  :muscle:
Your project meets most of the requirements :clap: :clap: :clap:  
You still need to fix some issued before you can go ahead:

- [x] Fix rubocop offenses
- [x] In your README file, add a screenshot
- [x] In your README file, add how to set up the project
- [x] #my_each when a block is given when self is a range returns the range itself after calling the given block once for each element in 
self
        `[1,2,3].my_each(&proc{|x| x>2}) == [1,2,3]`
- [x] #my_each_with_index when a block is given when self is a range returns the range itself after calling the given block once for each 
element in self
        `(1..3).my_each_with_index(&proc{|x| x>2}) == (1..3)`
- [x] #my_select returns an array containing all elements of enum for which the given block returns a true value::range
    `(1..3).select(&proc{|x| x>2}) == (1..3).my_select(&proc{|x| x>2})`
- [x] #my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
      `[5,5,5].all?(5) == [5,5,5].my_all?(5)`
- [x] #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern      
      `[5,5,[5]].all?(5) == [5,5,[5]].my_all?(5)`
- [x] #my_any? returns false if the block ever returns a value false or nil::range
    `(1..3).any?(&proc{|x| x==0}) == (1..3).my_any?(&proc{|x| x==0})`
- [x] #my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
      `['dog','car'].any?('cat') == ['dog','car'].my_any?('cat')`
- [x] #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern
      `['cat', 'dog','car'].any?('cat') == ['cat', 'dog','car'].my_any?('cat')`
- [x] #my_none? returns false if the block ever returns true for all elements::range
    `(1..3).none?(&proc{|num| num%2==0}) == (1..3).my_none?(&proc{|num| num%2==0})`
- [x] #my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern       
      `['dog','car'].none?(5) == ['dog','car'].my_none?(5)`
- [x] #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern       
      `[5,'dog','car'].none?(5) == [5,'dog','car'].my_none?(5)`
- [x] #my_map returns a new array with the results of running a given block once for every element in enum::range
    `(1..3).map(&proc{|num| num+1}) == (1..3).my_map(&proc{|num| num+1})`
- [x] #my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
       `[1,2,3].my_map(proc{|x|x%2}){|a|a*2} == [1,2,3].my_map(proc{|x|x%2})`
- [x] #my_inject raises a 'LocalJumpError' when no block or argument is given
    ```
    begin
      [1,2,3,4].my_inject
    rescue LocalJumpError
      true
    end
    ```


Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

ahmetbozaci committed:
fixed my_each for range

ahmetbozaci committed:
fixed my_inject for  if no block or argument given

ahmetbozaci committed:
added screenshot

ahmetbozaci committed:
fix my_all if given number or string argument

ahmetbozaci committed:
fixed my_any if given string or number

ahmetbozaci committed:
fixed my_none if given number or string argument

ahmetbozaci committed:
fixed my_map for proc given

ahmetbozaci commented:
Dear TSE,

1. I didn't have any rubocop errors in my first pull request. But after all changes ı checked again.
2. I think no need screenshot for this project but ı added screenshot.
3. I added "how to setup project" in readme file.
4. I fixed **my_each** method for range. **my_each_with_index** method fixed automatically after that.
5. I fixed **my_none, my_all,  my_any** methods if given argument as a string or number. Range errors fixed automatically because of my_each method.
6. I fixed **my_map** method for proc.
7. I already added statement if no block or argument given in my_inject method. I just missed "," after "LocalJumpError" so it fixed too.

SpaYco commented on enumerables.rb:
```diff
@@ -0,0 +1,263 @@
+    my_each do |i|
+      yield i, to_a.index(i)
+    end
+  end
```
- [x] when a block is given when self is an array it calls the given block once for each element in self

```Ruby
  my_each_with_index_output = ''
  block = proc { |num, idx| my_each_with_index_output += "Num: #{num}, idx: #{idx}\n" }
  enum.my_each_with_index(&block)
  # should be "Num: 1, idx: 0\nNum: 6, idx: 1\nNum: 1, idx: 2\nNum: 0, idx: 3\nNum: 0, idx: 4\nNum: 3, idx: 5\nNum:...: 5, idx: 94\nNum: 4, idx: 95\nNum: 2, idx: 96\nNum: 8, idx: 97\nNum: 0, idx: 98\nNum: 2, idx: 99\n"
  # but got "Num: 1, idx: 0\nNum: 6, idx: 1\nNum: 1, idx: 0\nNum: 0, idx: 3\nNum: 0, idx: 3\nNum: 3, idx: 5\nNum:...Num: 5, idx: 19\nNum: 4, idx: 13\nNum: 2, idx: 7\nNum: 8, idx: 11\nNum: 0, idx: 3\nNum: 2, idx: 7\n" instead
```

SpaYco commented on enumerables.rb:
```diff
@@ -0,0 +1,263 @@
+      end
+      true
+    end
+  end
```
- [x] when a Regex is passed as an argument it returns true if all of the collection matches the Regex

```Ruby
  words = %w[dog door rod blade]
  words.my_all?(/d/)
  # should be true but got false instead
```

SpaYco commented on enumerables.rb:
```diff
@@ -0,0 +1,263 @@
+      end
+      false
+    end
+  end
```
- [x] when a Regex is passed as an argument it returns true if all of the collection matches the Regex

```Ruby
  words = %w[dog door rod blade]
  words.my_any?(/d/)
  # should be true but got false instead
```

SpaYco commented on enumerables.rb:
```diff
@@ -0,0 +1,263 @@
+      end
+      true
+    end
+  end
```
- [x] when a Regex is passed as an argument it returns false if all of the collection matches the Regex

```Ruby
  words = %w[dog door rod blade]
  words.my_none?(/d/)
  # should be false but got true instead
```

SpaYco reviewed (changes requested):
# Changes Requested 

Hello! 👋 

Awesome job so far!! 👍 you still need to make some changes though! 

> I think no need screenshot for this project but ı added screenshot.

you're right, you don't need a screenshot, you can safely remove it.

## Required Changes ♻️ 

- [x] check the comments below 👇 

### please make the appropriate changes and ask for another code review ✔️ 

Happy Coding!

Best, 
Aziz Mejri 
_contact me on [twitter](https://twitter.com/iSpaYco) or slack (@SpaYco) if you need any help_

ahmetbozaci committed:
fixed my_any my_none my_all for regexp

ahmetbozaci commented:
Dear TSE,

1. I fixed **my_any, my_all , my_none** for regexp. Outputs given what should be.
2. I missed return "self" in my_each_with_index. I added. But I don't understand review. I test my code with a similar example as yours and it is working well.

voscarmv commented on enumerables.rb:
```diff
@@ -0,0 +1,264 @@
+    self
+  end
+
+  # my_each_with_index
+  def my_each_with_index()
+    return to_enum(:my_each_with_index) unless block_given?
+
+    my_each do |i|
+      yield i, to_a.index(i)
+    end
+    self
+  end
+
```
When testing this, the following error occurs:

```
  1) enumerables #my_each_with_index when a block is given when self is an array behaves like #each_with_index calls the given block once for each element in self
     Failure/Error: expect(my_each_with_index_output).to eq(each_with_index_output)
     
       expected: "Num: 4, idx: 0\nNum: 4, idx: 1\nNum: 7, idx: 2\nNum: 2, idx: 3\nNum: 8, idx: 4\nNum: 8, idx: 5\nNum:...: 1, idx: 94\nNum: 4, idx: 95\nNum: 7, idx: 96\nNum: 4, idx: 97\nNum: 7, idx: 98\nNum: 6, idx: 99\n"
            got: "Num: 4, idx: 0\nNum: 4, idx: 0\nNum: 7, idx: 2\nNum: 2, idx: 3\nNum: 8, idx: 4\nNum: 8, idx: 4\nNum:...\nNum: 1, idx: 16\nNum: 4, idx: 0\nNum: 7, idx: 2\nNum: 4, idx: 0\nNum: 7, idx: 2\nNum: 6, idx: 6\n"
     
       (compared using ==)
     
       Diff:
       
       
       
       
       
       @@ -1,101 +1,101 @@
        Num: 4, idx: 0
       -Num: 4, idx: 1
       +Num: 4, idx: 0
        Num: 7, idx: 2
        Num: 2, idx: 3
        Num: 8, idx: 4
       -Num: 8, idx: 5
       +Num: 8, idx: 4
        Num: 6, idx: 6
       -Num: 2, idx: 7
       -Num: 7, idx: 8
       -Num: 7, idx: 9
       -Num: 6, idx: 10
       +Num: 2, idx: 3
       +Num: 7, idx: 2
       +Num: 7, idx: 2
       +Num: 6, idx: 6
        Num: 3, idx: 11
       -Num: 8, idx: 12
       -Num: 8, idx: 13
       -Num: 2, idx: 14
       +Num: 8, idx: 4
       +Num: 8, idx: 4
       +Num: 2, idx: 3
        Num: 0, idx: 15
        Num: 1, idx: 16
       -Num: 3, idx: 17
       -Num: 4, idx: 18
       -Num: 1, idx: 19
       -Num: 1, idx: 20
       -Num: 8, idx: 21
       -Num: 7, idx: 22
       +Num: 3, idx: 11
       +Num: 4, idx: 0
       +Num: 1, idx: 16
       +Num: 1, idx: 16
       +Num: 8, idx: 4
       +Num: 7, idx: 2
        Num: 5, idx: 23
       -Num: 8, idx: 24
       -Num: 5, idx: 25
       -Num: 5, idx: 26
       -Num: 2, idx: 27
       -Num: 0, idx: 28
       -Num: 5, idx: 29
       -Num: 0, idx: 30
       -Num: 2, idx: 31
       -Num: 8, idx: 32
       -Num: 5, idx: 33
       -Num: 6, idx: 34
       -Num: 6, idx: 35
       -Num: 0, idx: 36
       -Num: 5, idx: 37
       -Num: 0, idx: 38
       -Num: 2, idx: 39
       -Num: 0, idx: 40
       -Num: 1, idx: 41
       -Num: 6, idx: 42
       -Num: 6, idx: 43
       -Num: 6, idx: 44
       -Num: 7, idx: 45
       -Num: 4, idx: 46
       -Num: 0, idx: 47
       -Num: 1, idx: 48
       -Num: 8, idx: 49
       -Num: 5, idx: 50
       -Num: 8, idx: 51
       -Num: 8, idx: 52
       -Num: 5, idx: 53
       -Num: 5, idx: 54
       -Num: 5, idx: 55
       -Num: 4, idx: 56
       -Num: 3, idx: 57
       -Num: 7, idx: 58
       -Num: 6, idx: 59
       -Num: 3, idx: 60
       -Num: 2, idx: 61
       -Num: 8, idx: 62
       -Num: 0, idx: 63
       -Num: 2, idx: 64
       -Num: 2, idx: 65
       -Num: 8, idx: 66
       -Num: 1, idx: 67
       -Num: 0, idx: 68
       -Num: 4, idx: 69
       -Num: 0, idx: 70
       -Num: 7, idx: 71
       -Num: 6, idx: 72
       -Num: 1, idx: 73
       -Num: 1, idx: 74
       -Num: 6, idx: 75
       -Num: 2, idx: 76
       -Num: 1, idx: 77
       -Num: 4, idx: 78
       -Num: 8, idx: 79
       -Num: 1, idx: 80
       -Num: 2, idx: 81
       -Num: 8, idx: 82
       -Num: 2, idx: 83
       -Num: 5, idx: 84
       -Num: 3, idx: 85
       -Num: 3, idx: 86
       -Num: 0, idx: 87
       -Num: 6, idx: 88
       -Num: 0, idx: 89
       -Num: 2, idx: 90
       -Num: 6, idx: 91
       -Num: 1, idx: 92
       -Num: 6, idx: 93
       -Num: 1, idx: 94
       -Num: 4, idx: 95
       -Num: 7, idx: 96
       -Num: 4, idx: 97
       -Num: 7, idx: 98
       -Num: 6, idx: 99
       +Num: 8, idx: 4
       +Num: 5, idx: 23
       +Num: 5, idx: 23
       +Num: 2, idx: 3
       +Num: 0, idx: 15
       +Num: 5, idx: 23
       +Num: 0, idx: 15
       +Num: 2, idx: 3
       +Num: 8, idx: 4
       +Num: 5, idx: 23
       +Num: 6, idx: 6
       +Num: 6, idx: 6
       +Num: 0, idx: 15
       +Num: 5, idx: 23
       +Num: 0, idx: 15
       +Num: 2, idx: 3
       +Num: 0, idx: 15
       +Num: 1, idx: 16
       +Num: 6, idx: 6
       +Num: 6, idx: 6
       +Num: 6, idx: 6
       +Num: 7, idx: 2
       +Num: 4, idx: 0
       +Num: 0, idx: 15
       +Num: 1, idx: 16
       +Num: 8, idx: 4
       +Num: 5, idx: 23
       +Num: 8, idx: 4
       +Num: 8, idx: 4
       +Num: 5, idx: 23
       +Num: 5, idx: 23
       +Num: 5, idx: 23
       +Num: 4, idx: 0
       +Num: 3, idx: 11
       +Num: 7, idx: 2
       +Num: 6, idx: 6
       +Num: 3, idx: 11
       +Num: 2, idx: 3
       +Num: 8, idx: 4
       +Num: 0, idx: 15
       +Num: 2, idx: 3
       +Num: 2, idx: 3
       +Num: 8, idx: 4
       +Num: 1, idx: 16
       +Num: 0, idx: 15
       +Num: 4, idx: 0
       +Num: 0, idx: 15
       +Num: 7, idx: 2
       +Num: 6, idx: 6
       +Num: 1, idx: 16
       +Num: 1, idx: 16
       +Num: 6, idx: 6
       +Num: 2, idx: 3
       +Num: 1, idx: 16
       +Num: 4, idx: 0
       +Num: 8, idx: 4
       +Num: 1, idx: 16
       +Num: 2, idx: 3
       +Num: 8, idx: 4
       +Num: 2, idx: 3
       +Num: 5, idx: 23
       +Num: 3, idx: 11
       +Num: 3, idx: 11
       +Num: 0, idx: 15
       +Num: 6, idx: 6
       +Num: 0, idx: 15
       +Num: 2, idx: 3
       +Num: 6, idx: 6
       +Num: 1, idx: 16
       +Num: 6, idx: 6
       +Num: 1, idx: 16
       +Num: 4, idx: 0
       +Num: 7, idx: 2
       +Num: 4, idx: 0
       +Num: 7, idx: 2
       +Num: 6, idx: 6
```

There seems to be a problem with the `idx` index being miscalculated. Judging from the `diff` part of the error message, the `to_a.index(i)` (line 19) is returning only the index of the first occurrenc of the element `i`. The index is supposed to be a consecutive value, starting at 0. Your code is trying to find only the index of the first occurrence of the value `i`. Therefor, if there is more than one of any single value in the array, it will return only the index of the first occurrence, instead of a consecutive series of indices. Please fix this.

You may test your code by checking that this produces the desired output:
```
        enum = [1, 1, 1, 2, 2, 2, 3, 3, 3]
        my_each_with_index_output = ''
        block = proc { |num, idx| my_each_with_index_output += "Num: #{num}, idx: #{idx}\n" }
        enum.each_with_index(&block)
        each_with_index_output = my_each_with_index_output.dup
        my_each_with_index_output = ''
        enum.my_each_with_index(&block)
        expect(my_each_with_index_output).to eq(each_with_index_output)
```

This should output a series with consecutive indices, and not with indices `[0, 0, 0, 1, 1, 1, 2, 2, 2]`

voscarmv reviewed (changes requested):
# Impressive work, team! 🖥️ 👀 ✨ 

- Congrats on fixing almost all of the failing methods!

## However, some changes are still required... 🧰 🔨 🔧 

- The one remaining method producing errors is `my_each_with_index`. Please check my code comments on how to fix it!

### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

ahmetbozaci committed:
fixed my_each_with_index

ahmetbozaci commented:
Fixed my_each_with_index method. Output given what should be for every array.

Srjnnnn reviewed (changes requested):
### ### Status: Changes Required :recycle: :stop_sign: 
Hello Ahmet. It's a pleasure to review your code and I think you've done a great job till now. :100: 
But your my_each_with_index doesn't work as expected.

- [x] enumerables #my_each_with_index when a block is given when self is a range returns the range itself after calling the given block once for each element in self. To pass this test please create a range variable and give it to your method and arrange your method's behaviour as each_with_index.

- [x]  enumerables #my_each_with_index when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self. It's pretty self-explanatory. Your method needs to call the block for once for each element in the block as each_with_index does.

- [x]   enumerables #my_each_with_index when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self. Please also try to do it with a hash variable.

### A general comment:
As I said earlier, you have done a great job. And you need to improve just one method and then it'll be done. :muscle: 
You just need to create a range and hash and and some blocks with them when calling the methods and make some experiments with them. After that, you'll find the exact behavior of the method. Try to mimic each_with_index. :hindu_temple: 

You can use previous TSE's example to work on your test cases.

Once you finish with the necessary changes, please submit for a re-review.
Have a nice day. 
If you have a question about the review, you can reach me via Slack.
Eyüp Sercan UYGUR.

ahmetbozaci committed:
fixed my_each_with_index for range and hash

ahmetbozaci commented:
Dear TSE,

- I fixed **my_each_with_index** for **range** and **hash** classes. İt is not giving error for range and hash classes now.

andresporras3423 reviewed (changes requested):
### Status: Changes Required :recycle:
**Hello Ahmet Bozacı** 
Nice job with the improvements to your project :muscle:  :muscle:
It passes all the automatic tests for the Enumerable methods :clap: :clap: :clap:  
We still need to fix some problems with rubocop:
- [x] You project still shows some rubocop offenses, check out the next link:
https://drive.google.com/file/d/1Tm7fzkmNxJJ1O67QP2QmEdfwUKAv0mKU/view
- [x] Don't overuse rubocop disabling
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

ahmetbozaci commented:
Dear TSE,

I just disabled these rubocop's features:
1. ModuleLength,
2. CyclomaticComplexity
3. PerceivedComplexity
4. MethodLength

Even I enabled them I can't see errors what you mention.
https://drive.google.com/file/d/1jFgWfJXooD-S8NhV2uHWcSaps5sUDv1k/view?usp=sharing

mahmoud717 commented on enumerables.rb:
```diff
@@ -0,0 +1,264 @@
+  end
+end
+
+# rubocop:enable Metrics/ModuleLength, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Metrics/MethodLength
```
## optional # 
move this line to the end of the page

mahmoud717 reviewed (approved):
## Project approved  # 
## Hi @ahmetbozaci  # 
nice job with the requested changes, your project is approved.
you can go ahead and merge this pull request
**some tips for your next projects**
- try as much as you can not to disable rubocop linters especially method length.
- when you are asked to do a change like adding a screenshot of your project, just go along with it.
## Optional # 
- delete the test file before merging the pr
- check my inline comment

wish you the best of luck with your next projects
